drivers/gpio/gpio_silabs_siwx917.c

```diff
@@ -0,0 +1,418 @@
+/*
+ * Copyright (c) 2024 Silicon Laboratories Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT silabs_siwx917_gpio
+
+#include "sl_si91x_driver_gpio.h"
+#include "sl_status.h"
+
+#include <errno.h>
+#include <zephyr/sys/util.h>
+#include <zephyr/irq.h>
+
+/* Zephyr GPIO header must be included after driver, due to symbol conflicts
+ * for GPIO_INPUT and GPIO_OUTPUT between preprocessor macros in the Zephyr
+ * API and struct member register definitions for the SiWx917 device.
+ */
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/gpio/gpio_utils.h>
+
+#if CONFIG_GPIO_SILABS_SIWX917_COMMON_INIT_PRIORITY >= CONFIG_GPIO_INIT_PRIORITY
+#error CONFIG_GPIO_SILABS_SIWX917_COMMON_INIT_PRIORITY must be less than \
+	CONFIG_GPIO_INIT_PRIORITY.
+#endif
+
+#define MAX_PORT_COUNT  4
+#define INVALID_PORT    0xFF
+#define INTERRUPT_COUNT 8
+
+/* Types */
+struct gpio_siwx917_common_config {
+	EGPIO_Type *reg;
+};
+
+struct gpio_siwx917_port_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	const struct device *parent;
+	int port;
+	int hal_port;
+	bool ulp;
+};
+
+struct gpio_siwx917_common_data {
+	/* a list of all ports */
+	const struct device *ports[MAX_PORT_COUNT];
+	sl_gpio_t interrupts[INTERRUPT_COUNT];
+};
+
+struct gpio_siwx917_port_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	/* port ISR callback routine address */
+	sys_slist_t callbacks;
+};
+
+static const uint8_t gpio_pad[76] = {
+	0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 1,    2,    3,    4,    5,    6,    7,
+	0xFF, 0xFF, 8,    0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 25,
+	26,   27,   28,   29,   30,   9,    9,    9,    9,    0xFF, 0xFF, 0xFF, 0xFF,
+	0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 10,   11,   12,   13,   14,   15,
+	16,   17,   18,   19,   20,   21,   0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 22,
+	23,   24,   25,   26,   27,   28,   29,   30,   31,   32,   33};
```

Comment:
* I did 8 values per line (16 is too wide for the 100-character-width coding standard), but clang-format reflowed it to this. I don't want to fight the automation.
* 0xFF are pads that don't exist
* Pad 9 is somehow special and the docs just say "it's already been configured for another purpose, don't reconfigure it" (I assume the ROM bootloader has set it up for _something_)
* Agree on NWP pads and shipping it in DT, I'll see what I can do. I'll probably use the unused "0" value for host pads (all the other ones are NWP by default)

Reply:
@jhedberg do you confirm our github workflow is configured to reformat the code? I am not comfortable with this kind of automatisation (see rule #1 of ["Manifesto for Agile Software Development"](https://agilemanifesto.org/): _Individuals and interactions over processes and tools_).

Reply:
First: The PR does not reformat, it just gives you warnings about there being a diff. AIUI, clang-format is not required, but it's available as a tool, and it will nag on your PR if it finds deviations. In the end, the coding standard and human decision wins over the checking done by clang-format -- using comments to suppress clang-format warnings is also allowed.

See https://docs.zephyrproject.org/latest/contribute/guidelines.html#clang-format for info.

I'm in the camp where I prefer not needing to think about formatting, so running clang-format is preferred to manually indenting according to the standard. This array was an edge case where I _almost_ wanted to wrap it in `clang-format off` comments, but I decided against it. The point is now moot anyway, since the array is now in DT instead.